cmd/download_grants_gov_db/handler.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	"github.com/usdigitalresponse/grants-ingest/internal/log"
+)
+
+var logger log.Logger
+
+// ScheduledEvent represents the invocation event for this Lambda function
+type ScheduledEvent struct {
+	Timestamp time.Time `json:"timestamp"`
+}
+
+// grantsURL returns the download URL for the Grants.gov database export.
+func (e *ScheduledEvent) grantsURL() string {
+	return fmt.Sprintf("%s/extract/GrantsDBExtract%sv2.zip",
+		env.GrantsGovBaseURL,
+		e.Timestamp.Format("20060102"),
+	)
+}
+
+// destinationS3Key returns the S3 object key where the database export should be stored.
+func (e *ScheduledEvent) destinationS3Key() string {
+	return fmt.Sprintf("sources/%s/grants.gov/archive.zip", e.Timestamp.Format("2006/01/02"))
+}
+
+// handleWithConfig is a Lambda function handler that is called with the ScheduledEvent invocation
+// event. When invoked, it streams a Grants.gov database export (zip file) to S3.
+func handleWithConfig(cfg aws.Config, ctx context.Context, event ScheduledEvent) error {
+	uploader := manager.NewUploader(s3.NewFromConfig(cfg, func(o *s3.Options) {
+		o.UsePathStyle = env.UsePathStyleS3Opt
+	}))
+	logger := log.With(logger,
+		"db_date", event.Timestamp.Format("2006-01-02"),
+		"source", event.grantsURL(),
+		"destination_bucket", env.DestinationBucket,
+		"destination_key", event.destinationS3Key(),
+	)
+
+	log.Debug(logger, "Starting remote file download")
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, event.grantsURL(), nil)
+	if err != nil {
+		return log.Errorf(logger, "Error configuring download request for source archive", err)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return log.Errorf(logger, "Error initiating download request for source archive", err)
+	}
+	defer resp.Body.Close()
+	if err := validateDownloadResponse(resp); err != nil {
+		return log.Errorf(logger, "Error downloading source archive", err)
+	}
+	logger = log.With(logger, "source_size_bytes", resp.ContentLength)
```

Comment:
To expand on the previous explanation: This is doing the same thing, but this time, overwriting the `logger` that was scoped to this function starting on line 42. Just like before, we pass in the previous `logger` so that the prior configuration is preserved and updated with an additional `source_size_bytes` key/value pair. Doing so allows us to accumulate context progressively throughout the function.

Note that if we didn't want to have this particular key/value pair accumulate forever, we could either add it to every `log.Level(...)` call explicitly, or we could scope a new variable name (like `sourceSizeLogger` or something) if we needed to reuse it a few times.